### ISSUE TITLE
task subclass not respecting tenants

### ISSUE THREAD

**Issue by thenewguy** [maciej-gol/tenant-schemas-celery#43]
Should tasks using class based definitions work here?  My tasks using the [Jobtastic baseclass](https://github.com/PolicyStat/jobtastic/blob/master/jobtastic/task.py#L36) do not respect the tenants and always run against the public schema.

I don't see anything related that would cause this, so I am curious if this is expected to work or is a new feature?

**Comment by maciej-gol** [comment]
Will look into this, but AFAIK it should work out of the box as long as the
subclass calls super() properly.

Thanks!

wt., 30 cze 2020, 18:21 użytkownik thenewguy <notifications@github.com>
napisał:

**Comment by maciej-gol** [comment]
@thenewguy could you please show me an example how you integrate both this app and the jobtastic task?

**Comment by thenewguy** [comment]
@maciej-gol working on a test case now

**Comment by thenewguy** [comment]
I am looking through your test cases to see how you run the workers, but here is my test that fails running against celery's start_worker:

tasks:
```
@shared_task
def get_schema_name():
    return connection.schema_name

class JobtasticSchemaTask(JobtasticTask):
    significant_kwargs = []
    herd_avoidance_timeout = -1
    cache_duration = -1
    
    def calculate_result(self):
        return connection.schema_name
```

tests:
```
    def test_tenant_context_for_celery_shared_tasks(self):
        self.assertEqual(get_schema_name.delay().get(timeout=30), 'test')
        with public_schema_context():
            self.assertEqual(get_schema_name.delay().get(timeout=30), 'public')
        self.assertEqual(get_schema_name.delay().get(timeout=30), 'test')
    
    def test_tenant_context_for_jobtastic_tasks(self):
        self.assertEqual(JobtasticSchemaTask.delay().get(timeout=30), 'test')
        with public_schema_context():
            self.assertEqual(JobtasticSchemaTask.delay().get(timeout=30), 'public')
        self.assertEqual(JobtasticSchemaTask.delay().get(timeout=30), 'test')
```

Note that `test_tenant_context_for_celery_shared_tasks()` passes and `test_tenant_context_for_jobtastic_tasks()` fails with:
```
======================================================================
FAIL: test_tenant_context_for_jobtastic_tasks (integration.tests.test_workers.WorkersCeleryTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vagrant/integration/tests/test_workers.py", line 48, in test_tenant_context_for_jobtastic_tasks
    self.assertEqual(JobtasticSchemaTask.delay().get(timeout=30), 'test')
AssertionError: 'public' != 'test'
- public
+ test
```

**Comment by maciej-gol** [comment]
So the problem is that given:

```python
class JobtasticSchemaTask(JobtasticTask):
    significant_kwargs = []
    herd_avoidance_timeout = -1
    cache_duration = -1

    def calculate_result(self):
        return connection.schema_name

def do_smth():
    JobtasticSchemaTask.delay()
```

you don't actually call the tenant-specific implementation at all. In order to fix that, you need to add `TenantTask` as the parent class:

```python
class JobtasticSchemaTask(TenantTask, JobtasticTask):
    significant_kwargs = []
    herd_avoidance_timeout = -1
    cache_duration = -1

    def calculate_result(self):
        return connection.schema_name
```

Unfortunately, currently this app doesn't provide a proper mixin that implements the tenancy logic. Thankfully, Python correctly resolves diamond inheritance here.

**Comment by thenewguy** [comment]
@maciej-gol is there any way to make this happen transparently so that 3rd party app tasks can be used?

**Comment by maciej-gol** [comment]
Are you using the `CeleryApp` provided by this app as your default app?

**Comment by thenewguy** [comment]
Yes, per docs.

```
from __future__ import absolute_import, unicode_literals
import os

from celery.signals import setup_logging
from tenant_schemas_celery.app import CeleryApp as TenantAwareCeleryApp

# set the default Django settings module for the 'celery' program.
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'integration.settings')

app = TenantAwareCeleryApp('integration')

# Using a string here means the worker doesn't have to serialize
# the configuration object to child processes.
# - namespace='CELERY' means all celery-related configuration keys
#   should have a `CELERY_` prefix.
app.config_from_object('django.conf:settings', namespace='CELERY')

@setup_logging.connect
def config_loggers(*args, **kwags):
    from logging.config import dictConfig
    from django.conf import settings
    dictConfig(settings.LOGGING)

# Load task modules from all registered Django app configs.
app.autodiscover_tasks()

from celery.contrib.testing.tasks import ping
```

**Comment by maciej-gol** [comment]
:+1: I think I can work something out. But then, I'm afraid you will need to continue with the subclassing.

**Comment by thenewguy** [comment]
@maciej-gol What do you mean?  Can we make the test work in my PR as written?  Happy to do the work if you can give me some direction

**Comment by maciej-gol** [comment]
So I think it's possible to make this work automatically, by implementing our own `celery.app.registry:TaskRegistry` subclass and passing it's import path as `registry_cls` attribute of the `CeleryApp`.

The new registry, when it's `register()` is called, should check if the argument is a type and a subclass of our TenantMixin. If it's a type but not a subclass, we should dynamically create a new class that inherits from the TenantMixin and the argument, and register it appropriately.

If the argument is not a type (thus a task instance) I would really opt for least surprise here, and just continue without altering it.

So, to sum up, there are two tasks to do:

* The `TenantMixin` .
* The `TaskRegistry` subclass.

It's not a lot of work, but until it's done you will have to deal with the original subclassing workaround

**Comment by thenewguy** [comment]
@maciej-gol great! Let's see how I fare

**Comment by thenewguy** [comment]
Tasks that use the legacy `from celery.task import Task as LegacyTask` base class and also tasks that use the task decorator without inheritance are passing.

However, I am struggling with new style tasks:

```
from celery import shared_task, Task
from django.db import connection

class SchemaClassTask(Task):
    @property
    def connection_schema_name(self):
        return connection.schema_name


@shared_task(base=SchemaClassTask, bind=True)
def get_schema_from_class_task(self):
    return self.connection_schema_name
```

It seems like the [recommended way to write tests now per docs](https://docs.celeryproject.org/en/stable/userguide/tasks.html#task-inheritance) so I feel like it is important to support it.

**Comment by thenewguy** [comment]
They seem to be passed as functions.  You said you didn't want to touch those hence the quandary.  Otherwise, I think we could wrap the task passed to registration if it isn't an instance of TenantTask and proxy the calls we need?

**Comment by thenewguy** [comment]
@maciej-gol So, PR #45 seems to be working and logs indicate my tasks are running in the correct schema context now.  However, I am still getting some odd behavior on tenants where the tasks seem to lose the tenant during computation.

For example, a long running task starts by retrieving the record from the database on the tenant schema.  Then it operates for some time.  Finally it attempts to save a model with a foreign key pointing to the record it previously retrieved but errors:
```
django.db.utils.IntegrityError: insert or update on table "long_table_a" violates foreign key constraint "long_constraint_b"
DETAIL:  Key (record_id)=(2090833968) is not present in table "long_table_c"
```

The confusing part is that it starts by retrieving the record from the database.  So it must exist - and afterwards the record is still viewable on the tenant site.

Are you familiar with anything like this?  The only thing I can think of is that I need to be wrapping the tasks with a transaction before the tenant schema is set.  However, I am not sure how to hook that in with the `task_prerun` and `task_postrun` signals.

Initially I thought it was a pgbouncer issue.  I have removed pgbouncer from the equation and am still observing the issue.  All tasks run against the public schema successfully.  However, many of the tenant tasks suffer the problem above.  I have removed pgbouncer from the equation on the task workers and they are operating against the database directly.

**Comment by maciej-gol** [comment]
Can you print the schema name when the exception occurs? Also, what kind of
concurrency model you use in celery? multiprocessing/threads/gevent?

śr., 8 lip 2020 o 21:26 thenewguy <notifications@github.com> napisał(a):

**Comment by thenewguy** [comment]
@maciej-gol Yes, the schema name is included in the log.  For example:

```
ERROR [ tenant_3094fa0d2854419c9e58a2b1b126c922 | None | 2020-07-08 18:38:35,817Z | none | multimedia.tasks.GenericModelClassMethodTask | 959bd153-bf52-4ca0-8a55-3897b6bce907 ] multimedia.tasks: Encountered unhandled exception from class method encode_from_media_reference
Traceback (most recent call last):
  File "/var/lib/integration/django/db/backends/base/base.py", line 243, in _commit
    return self.connection.commit()
psycopg2.errors.ForeignKeyViolation: insert or update on table "adaptive_adaptivemediarendition" violates foreign key constraint "adaptive_adaptivemed_media_id_e1a7ac71_fk_multimedi"
DETAIL:  Key (media_id)=(2090833968) is not present in table "multimedia_enhancedmedia".
```

where `tenant_3094fa0d2854419c9e58a2b1b126c922` is the tenant schema (it would be "public" if it was the public schema) and `959bd153-bf52-4ca0-8a55-3897b6bce907` is the task id.

The media record `2090833968` it complains about was fetched at the start of the task a long time before the error is raised at commit.  And it can't process to output the other information in the logs without having the media record because it contains the file field.

For example, this log entry output by the worker during the task before the error:
```
INFO [ tenant_3094fa0d2854419c9e58a2b1b126c922 | None | 2020-07-08 18:36:28,696Z | none | multimedia.tasks.GenericModelClassMethodTask | 959bd153-bf52-4ca0-8a55-3897b6bce907 ] django.multimedia.editor: loaded file clip '/tmp/cae1f8475303411a92b1ca31ab0a5c18/source.flv' duration: '29.68'
```

Celery workers are running in the basic multiprocessing mode like:
```
celery worker --app=integration --loglevel=debug --events --queues=slow --prefetch-multiplier=1 -- worker.prefetch_multiplier=1
```

**Comment by maciej-gol** [comment]
@thenewguy I think the issue is unrelated to the change to task registration. Could be caused by the media object being long gone during the execution of the task it self.

Regarding the PR, very good fix, thanks a lot!
